File: ios/Packages/StructoKit/Sources/StructoKit/Streaks.swift
// Streak engine — Swift implementation of spec/streaks/rules.md.
// Must stay conformant with spec/streaks/vectors.json; the TypeScript
// engine (web) and Postgres compute_streak() are tested against the same
// file. Derived data: recompute, never merge (§11.5).

public struct StreakHabit: Sendable {
    public enum Kind: String, Sendable {
        case binary, quantity, frequency
    }

    public enum Schedule: Equatable, Sendable {
        case daily
        case weekdays([Int]) // ISO, 1 = Monday
        case perWeek
    }

    public var type: Kind
    public var schedule: Schedule
    public var targetValue: Double?
    public var timesPerWeek: Int?
    public var pauses: [(from: CivilDate, to: CivilDate)]

    public init(
        type: Kind,
        schedule: Schedule,
        targetValue: Double? = nil,
        timesPerWeek: Int? = nil,
        pauses: [(from: CivilDate, to: CivilDate)] = []
    ) {
        self.type = type
        self.schedule = schedule
        self.targetValue = targetValue
        self.timesPerWeek = timesPerWeek
        self.pauses = pauses
    }
}

public struct StreakResult: Equatable, Sendable {
    public enum Unit: String, Sendable {
        case days, weeks
    }

    public var current: Int
    public var best: Int
    public var unit: Unit

    public init(current: Int, best: Int, unit: Unit) {
        self.current = current
        self.best = best
        self.unit = unit
    }
}

public enum Streaks {
    /// - Parameter logs: (loggedFor, value) pairs; negative values are
    ///   compensating entries (§5.4). Order is irrelevant.
    public static func compute(
        habit: StreakHabit,
        logs: [(day: CivilDate, value: Double)],
        today: CivilDate,
        weekStart: Int = 1
    ) -> StreakResult {
        let unit: StreakResult.Unit = habit.type == .frequency ? .weeks : .days

        // Day-level completion: net total per day reaches the threshold.
        var totals: [Int: Double] = [:]
        for log in logs {
            totals[log.day.dayNumber, default: 0] += log.value
        }
        let threshold = habit.type == .quantity ? (habit.targetValue ?? 1) : 1
        let completed = Set(totals.filter { $0.value >= threshold }.keys)
        let todayN = today.dayNumber
        guard let minDay = completed.min(), minDay <= todayN else {
            return StreakResult(current: 0, best: 0, unit: unit)
        }
        let paused = { (d: Int) -> Bool in
            habit.pauses.contains { d >= $0.from.dayNumber && d <= $0.to.dayNumber }
        }
        let scheduled = { (d: Int) -> Bool in
            if case .weekdays(let days) = habit.schedule {
                return days.contains(CivilDate(dayNumber: d).isoWeekday)
            }
            return true
        }

        var current = 0
        var best = 0
        var run = 0

        switch habit.type {
        case .binary, .quantity:
            // current: backward walk; unlogged scheduled *today* is pending
            var d = todayN
            while d >= minDay {
                if scheduled(d) && !paused(d) {
                    if completed.contains(d) {
                        current += 1
                    } else if d != todayN {
                        break
                    }
                }
                d -= 1
            }
            // best: forward scan with the same neutrality rules
            for day in minDay...todayN {
                if scheduled(day) && !paused(day) {
                    if completed.contains(day) {
                        run += 1
                        best = max(best, run)
                    } else if day != todayN {
                        run = 0
                    }
                }
            }
            return StreakResult(current: current, best: max(best, current), unit: unit)

        case .frequency:
            let timesPerWeek = habit.timesPerWeek ?? 1
            let weekOf = { (d: Int) -> Int in
                let dow = CivilDate(dayNumber: d).isoWeekday
                return d - (((dow - weekStart) + 7) % 7)
            }
            let todayWeek = weekOf(todayN)
            let firstWeek = weekOf(minDay)

            struct WeekInfo {
                var neutral: Bool
                var satisfied: Bool
            }
            let info = { (w: Int) -> WeekInfo in
                var pausedDays = 0
                var count = 0
                for d in w..<(w + 7) {
                    if paused(d) { pausedDays += 1 }
                    if completed.contains(d) { count += 1 }
                }
                let active = 7 - pausedDays
                if active == 0 { return WeekInfo(neutral: true, satisfied: false) }
                let needed = (timesPerWeek * active + 6) / 7 // ceil
                return WeekInfo(neutral: false, satisfied: count >= needed)
            }

            var w = todayWeek
            while w >= firstWeek {
                let i = info(w)
                if !i.neutral {
                    if i.satisfied {
                        current += 1
                    } else if w != todayWeek {
                        break
                    }
                }
                w -= 7
            }
            w = firstWeek
            while w <= todayWeek {
                let i = info(w)
                if !i.neutral {
                    if i.satisfied {
                        run += 1
                        best = max(best, run)
                    } else if w != todayWeek {
                        run = 0
                    }
                }
                w += 7
            }
            return StreakResult(current: current, best: max(best, current), unit: unit)
        }
    }
}

File: web/src/lib/streaks.ts
// Habit streak engine — TypeScript implementation of spec/streaks/rules.md.
// Must stay conformant with spec/streaks/vectors.json (streaks.test.ts runs
// every vector in CI); Postgres implements the same rules in compute_streak()
// (0005) for the nightly habit_stats rollup. Client-side computation gives
// instant optimistic streaks; the server value wins on drift (recompute,
// never merge — §11.5).

export type StreakHabit = {
  type: "binary" | "quantity" | "frequency";
  schedule: { kind: "daily" } | { kind: "weekdays"; days: number[] } | { kind: "per_week" };
  target_value?: number | null;
  times_per_week?: number | null;
  pauses?: Array<{ from: string; to: string }>;
};

export type StreakResult = {
  current: number;
  best: number;
  unit: "days" | "weeks";
};

type LogPair = [string, number]; // [logged_for YYYY-MM-DD, value]

const DAY_MS = 86_400_000;

function ymdToMs(ymd: string): number {
  const [y, m, d] = ymd.split("-").map(Number);
  return Date.UTC(y, m - 1, d);
}

function msToYmd(ms: number): string {
  const d = new Date(ms);
  return `${d.getUTCFullYear()}-${String(d.getUTCMonth() + 1).padStart(2, "0")}-${String(d.getUTCDate()).padStart(2, "0")}`;
}

function isodow(ms: number): number {
  const dow = new Date(ms).getUTCDay();
  return dow === 0 ? 7 : dow;
}

/** Start-of-week (ms) containing `ms`, weeks bounded by week_start. */
function weekOf(ms: number, weekStart: number): number {
  return ms - ((isodow(ms) - weekStart + 7) % 7) * DAY_MS;
}

function isPaused(pauses: Array<{ from: string; to: string }>, ms: number): boolean {
  return pauses.some((p) => ms >= ymdToMs(p.from) && ms <= ymdToMs(p.to));
}

function isScheduled(habit: StreakHabit, ms: number): boolean {
  return habit.schedule.kind === "weekdays"
    ? habit.schedule.days.includes(isodow(ms))
    : true;
}

export function computeStreak(
  habit: StreakHabit,
  logs: LogPair[],
  today: string,
  weekStart = 1,
): StreakResult {
  const unit: StreakResult["unit"] = habit.type === "frequency" ? "weeks" : "days";
  const pauses = habit.pauses ?? [];

  // Day-level completion: net total per logged_for reaches the threshold.
  const totals = new Map<string, number>();
  for (const [day, value] of logs) {
    totals.set(day, (totals.get(day) ?? 0) + value);
  }
  const threshold = habit.type === "quantity" ? Number(habit.target_value ?? 1) : 1;
  const completed = new Set<number>();
  for (const [day, total] of totals) {
    if (total >= threshold) completed.add(ymdToMs(day));
  }
  if (completed.size === 0) return { current: 0, best: 0, unit };

  const todayMs = ymdToMs(today);
  const minMs = Math.min(...completed);
  let current = 0;
  let best = 0;
  let run = 0;

  if (habit.type === "binary" || habit.type === "quantity") {
    // current: backward walk; an unlogged scheduled *today* is pending
    for (let d = todayMs; d >= minMs; d -= DAY_MS) {
      if (isScheduled(habit, d) && !isPaused(pauses, d)) {
        if (completed.has(d)) current++;
        else if (d !== todayMs) break;
      }
    }
    // best: forward scan with the same neutrality rules
    for (let d = minMs; d <= todayMs; d += DAY_MS) {
      if (isScheduled(habit, d) && !isPaused(pauses, d)) {
        if (completed.has(d)) {
          run++;
          best = Math.max(best, run);
        } else if (d !== todayMs) {
          run = 0;
        }
      }
    }
    return { current, best: Math.max(best, current), unit };
  }

  // frequency: weeks satisfied when completed days >= pro-rata threshold
  const timesPerWeek = Number(habit.times_per_week ?? 1);
  const todayWeek = weekOf(todayMs, weekStart);
  const firstWeek = weekOf(minMs, weekStart);

  const weekInfo = (w: number) => {
    let pausedDays = 0;
    let count = 0;
    for (let d = w; d < w + 7 * DAY_MS; d += DAY_MS) {
      if (isPaused(pauses, d)) pausedDays++;
      if (completed.has(d)) count++;
    }
    const activeDays = 7 - pausedDays;
    return {
      neutral: activeDays === 0,
      satisfied:
        activeDays > 0 && count >= Math.ceil((timesPerWeek * activeDays) / 7),
    };
  };

  for (let w = todayWeek; w >= firstWeek; w -= 7 * DAY_MS) {
    const info = weekInfo(w);
    if (info.neutral) continue;
    if (info.satisfied) current++;
    else if (w !== todayWeek) break;
  }
  for (let w = firstWeek; w <= todayWeek; w += 7 * DAY_MS) {
    const info = weekInfo(w);
    if (info.neutral) continue;
    if (info.satisfied) {
      run++;
      best = Math.max(best, run);
    } else if (w !== todayWeek) {
      run = 0;
    }
  }
  return { current, best: Math.max(best, current), unit };
}

/** Net logged total for one user-local day (compensating entries included). */
export function dayTotal(logs: LogPair[], day: string): number {
  return logs.reduce((sum, [d, v]) => (d === day ? sum + v : sum), 0);
}

export { msToYmd, ymdToMs, isScheduled, DAY_MS };
